### DISCUSSION TITLE
Visual Studio SSL not working

### DISCUSSION THREAD

**Discussion by DeuSCaNoN** [uNetworking/uWebSockets#1451] [Q&A] [answered]
I suspect I'm doing some configuration wrong but I have this code
```		
uWS::SocketContextOptions sslContext{
			"privkey.pem",
			"fullchain.pem",
			"1234",
			"",
			"",
			0
};
```
being piped into my uWS::SSLApp but whenever I try to connect to it using https://localhost or wss://localhost nothing seems to load. 

using http:// and ws:// works though so it seems the routes just aren't being configured with SSL.

I just installed uwebsockets through vcpkg and just included the .h file. Is there some kind of preprocessor flag I need to set so that SSL works and isn't silently failing?

**Comment by Jacob-Burckhardt** [comment]
Check for error messages from the client.  The client might be saying something like that the certificate's common name doesn't match.  If you have trouble finding the client logs, try using this client instead:

gnutls-cli -p 443 localhost

It should say what the common name is.  Then rerun that command using a domain name that matches the common name.  If that works, then run the client you originally tried to use with that domain name instead of localhost.

Now that you are using a different domain name in your client, ensure that the different domain name still refers to your uWebSockets server.  If it does not refer to that, set up DNS to fix that or use /etc/hosts as a quick temporary workaround.

**Reply by DeuSCaNoN** [reply]
I set have 2 apps that uses the same ContextOptions, i listening on port 80 and 1 listening on port 18080. I get this when i run
gnutls-cli -p 80 localhost
gnutls-cli -p 18080 localhost

`*** Fatal error: The TLS connection was non-properly terminated.`

Just a note, I am developing on Windows using Visual Studios Community 2017, just wondering if I'm doing something wrong because I saw a openstack answer that said uWebsockets can be built without SSL and that can cause TLS to fail silently.

**Reply by Jacob-Burckhardt** [reply]
I haven't used uWebSockets with Windows or Visual Studios.  I use Linux.  It looks like defining this preprocessor flag is required for both OSs to use TLS:

WITH_OPENSSL=1

Have you defined that?

**Reply by DeuSCaNoN** [reply]
I'm not 100% sure how to do that in Visual Studios, is that preprocessor definition i have to define or a command line argument i have to add to the Msbuild.exe call?

**Reply by Jacob-Burckhardt** [reply]
This link seems to say how to do that:

https://docs.microsoft.com/en-us/cpp/build/reference/d-preprocessor-definitions?view=msvc-170

**Reply by DeuSCaNoN** [reply]
I just added it in and it still had problems. I also confirmed in the build command line it added this
` /D "WITH_OPENSSL"` so im pretty sure im compiling with it now

**Reply by Jacob-Burckhardt** [reply]
I rechecked the uWebSockets source code and I see WITH_OPENSSL is not checked by the compiler.  It is checked by makefiles that work on Linux and Mac.  So you might as well undo that /D "WITH_OPENSSL" change.

I recommend you search the web for how to enable SSL for uWebSockets using Visual Studio or vcpkg.

**Reply by DeuSCaNoN** [reply]
ok thanks, i may have to move onto another library then that works better on windows.

**Comment by ghost** [comment, marked as answer]
You need to build uSockets with SSL support, otherwise it will just default to regular TCP (that's a core feature - same business logic without any changes run on both TCP and TLS).

~https://github.com/microsoft/vcpkg/blob/master/ports/usockets/CMakeLists.txt#L25~

~Looks like vcpkg has CMAKE_USE_OPENSSL you need to set during install.~

~If you compile uSockets yourself, without vcpkg, it is LIBUS_USE_OPENSSL~

**Reply by ghost** [reply]
Scratch that. It is even easier than so!

vcpkg even has the concept of "features" and you just enable the features according to docs in vcpkg. ssl is a feature, so I guess you do vcpkg install usockets[ssl]

![screencapture-vcpkg-io-en-packages-html-2022-05-11-04_58_35](https://user-images.githubusercontent.com/3397140/167760082-645c0d17-eddc-447e-aedc-b4884e6fc53f.png)

**Reply by DeuSCaNoN** [reply]
THANK YOU SO MUCH!
@alexhultman just one further question, I only installed uwebsockets through vcpkg and it's already compiling, If I also install uSockets through vcpkg how does the original uwebsockets library know to link my new uSockets libs with ssl compiled?

Edit: nevermind, i figured it out, uninstall all my vcpkg packages, reinstall usockets with the ssl feature then install uwebsockets and it works
